File: leetcode/algorithms/257_binary_tree_paths/main.dart
class TreeNode {
  int val;
  TreeNode? left;
  TreeNode? right;
  TreeNode([this.val = 0, this.left, this.right]);
}

class BinaryTreePaths {
  /**
   * DFS + Backtracking
   *
   * Complexities:
   *   N - The number of nodes in `root`
   *   H - The height of tree in `root`
   *   - Time Complexity: O(N * H)
   *   - Space Complexity: O(H)
   */
  void calculatePath(TreeNode? root, List<String> stringArray, String string) {
    if (root != null) {
      string = "$string${string.isEmpty ? "" : "->"}${root.val}";

      if (root.left == null && root.right == null) {
        stringArray.add(string);
      } else {
        calculatePath(root.left, stringArray, string);
        calculatePath(root.right, stringArray, string);
      }
    }
  }

  List<String> binaryTreePaths(TreeNode? root) {
    List<String> stringArray = List<String>.empty(growable: true);

    calculatePath(root, stringArray, "");

    return stringArray;
  }


  // Solution
  /**
   * DFS + Backtracking
   *
   * Complexities:
   *   N - The number of nodes in `root`
   *   H - The height of tree in `root`
   *   - Time Complexity: O(N * H)
   *   - Space Complexity: O(H)
   */
  List<String> solution(TreeNode? root) {
    List<String> result = [];
    if (root == null) {
      return result;
    }

    List<int> currentPath = [];
    _backtrack(root, currentPath, result);

    return result;
  }

  void _backtrack(TreeNode node, List<int> currentPath, List<String> result) {
    currentPath.add(node.val);

    if (node.left == null && node.right == null) {
      result.add(currentPath.join("->"));
    } else {
      if (node.left != null) {
        _backtrack(node.left!, currentPath, result);
      }
      if (node.right != null) {
        _backtrack(node.right!, currentPath, result);
      }
    }

    currentPath.removeLast();
  }
}

File: leetcode/algorithms/257_binary_tree_paths/main.js
function TreeNode(val, left, right) {
  this.val = val === undefined ? 0 : val;
  this.left = left === undefined ? null : left;
  this.right = right === undefined ? null : right;
}

/**
 * DFS + Backtracking
 *
 * Complexities:
 *   N - The number of nodes in `root`
 *   H - The height of tree in `root`
 *   - Time Complexity: O(N * H)
 *   - Space Complexity: O(H)
 */
/**
 * @param {TreeNode} root
 * @return {string[]}
 */
var binaryTreePaths = function (root) {
  if (!root) {
    return [];
  }

  if (!root.left && !root.right) {
    return [`${root.val}`];
  }

  const leftArray = binaryTreePaths(root.left);
  const rightArray = binaryTreePaths(root.right);

  const left = leftArray.map((v_) => `${root.val}->${v_}`);
  const right = rightArray.map((v_) => `${root.val}->${v_}`);

  return [...left, ...right];
};


// Solution
/**
 * DFS + Backtracking
 *
 * Complexities:
 *   N - The number of nodes in `root`
 *   H - The height of tree in `root`
 *   - Time Complexity: O(N * H)
 *   - Space Complexity: O(H)
 */
var solution = function (root) {
  const result = [];

  function dfs(node, currentPath) {
    if (!node) {
      return;
    }

    const nextPath = currentPath
      ? `${currentPath}->${node.val}`
      : `${node.val}`;

    if (!node.left && !node.right) {
      result.push(nextPath);
      return;
    }

    dfs(node.left, nextPath);
    dfs(node.right, nextPath);
  }

  dfs(root, "");

  return result;
};
